CRM/Upgrade/Form.php

```diff
@@ -383,18 +396,51 @@ public function checkUpgradeableVersion($currentVer, $latestVer) {
 
     // check for mysql trigger privileges
     if (!\Civi::settings()->get('logging_no_trigger_permission') && !CRM_Core_DAO::checkTriggerViewPermission(FALSE, TRUE)) {
-      $error = ts('CiviCRM %1 requires MySQL trigger privileges.',
+      $errors[] = ts('CiviCRM %1 requires MySQL trigger privileges.',
         [1 => $latestVer]);
     }
 
     if (CRM_Core_DAO::getGlobalSetting('thread_stack', 0) < (1024 * self::MINIMUM_THREAD_STACK)) {
-      $error = ts('CiviCRM %1 requires MySQL thread stack >= %2k', [
+      $errors[] = ts('CiviCRM %1 requires MySQL thread stack >= %2k', [
         1 => $latestVer,
         2 => self::MINIMUM_THREAD_STACK,
       ]);
     }
 
-    return $error;
+    // CIVICRM_CRED_KEYS was introduced in 5.34. We make it required in 6.10+.
+    // Some sites may not have it (if installed before 5.34; or if manually disabled).
+    if (!defined('CIVICRM_CRED_KEYS') || in_array(trim(CIVICRM_CRED_KEYS), ['', 'plain'])) {
+
+      // Aaarg. We want the help-text to point you to the actual settings file, but it's not reliable.
+      // In 5.76+ on Standalone, CIVICRM_SETTINGS_PATH is sometimes overloaded as `civicrm.standalone.php`.
+      // We'll degrade gracefully.
+      $settingsPath = (basename(CIVICRM_SETTINGS_PATH) === 'civicrm.settings.php') ? CIVICRM_SETTINGS_PATH : NULL;
+
+      // Same generator formula as new installations (GenerateCredKey.civi-setup.php). Slightly oversizes to compensate for alphanumeric filter.
+      $newKey = 'aes-cbc:hkdf-sha256:' . preg_replace(';[^a-zA-Z0-9];', '', base64_encode(random_bytes(37)));
+
+      $code = [];
+      if ($settingsPath) {
+        $code[] = "// FILE: " . CIVICRM_SETTINGS_PATH;
+        $code[] = "";
+      }
+      $code[] = "if (!defined('CIVICRM_CRED_KEYS')) {";
+      $code[] = sprintf("  define('CIVICRM_CRED_KEYS', %s);", var_export($newKey, TRUE));
+      $code[] = "}";
+
+      $parts = [];
+      $parts[] = ts('CiviCRM %1 requires <code>CIVICRM_CRED_KEYS</code>. Please add it to <code>civicrm.settings.php</code>.', [
+        1 => $latestVer,
+      ]);
+      $parts[] = sprintf("<pre>\n%s\n</pre>", htmlentities(implode("\n", $code), ENT_COMPAT));
+      $parts[] = "(This example includes a randomly generated key. You may copy it directly or create your own random value.)";
```

Comment:
missing ts()

I think it's probably ok to not mention about rotating keys, since as discussed they likely don't have any working encrypted settings now if they were missing CRED or using "plain". And it links to the docs.

Reply:
Thanks, fixed the `ts()`.

I suspect that folks who left it to now probably won't care too much. FWIW, it links to the page on the Sysadmin Guide which does mention rotation...